sys/include/net/cord/lc.h

```diff
@@ -0,0 +1,181 @@
+    cord_lc_rd_t *rd;               /**< RD server the endpoint registered at */
+} cord_lc_ep_t;
+
+/**
+ * @brief  Information about a resource registered at a RD server
+ */
+typedef struct {
+    char path[NANOCOAP_URI_MAX];      /**< Resource path on the endpoint */
+    char rt[CORD_LC_ATTR_LEN];        /**< Application specific resource type */
+    char ifdesc[CORD_LC_ATTR_LEN];    /**< Interface description */
+    char anchor[NANOCOAP_URI_MAX];    /**< Anchor */
+} cord_lc_res_t;
+
+/**
+ * @brief Discover the lookup interface resources of the target RD server
+ *
+ * @param[out] rd       information about RD server at @p remote
+ *
+ * @return CORD_LC_OK on success
+ * @return CORD_LC_TIMEOUT if the discovery request times out
+ * @return CORD_LC_NORSC if one of the lookup interface for {res, ep, gp}
+ *                       resource type not found.
+ *                       Result **can** still be used
+ * @return CORD_LC_ERR on any other internal error
+ */
+int cord_lc_rd_init(cord_lc_rd_t *rd, const sock_udp_ep_t *remote);
+
+/**
+ * @brief Raw lookup for registered resources at a RD server
+ *
+ * @param[in]  rd               the target RD server
+ * @param[in]  content_format   data format wanted from RD server
+ * @param[out] result           buffer holding the raw response
+ * @param[in]  maxlen           max length of @p result
+ *
+ * @return length of result on success
+ * @return CORD_LC_NORSC if there is no resource lookup interface at @p rd
+ * @return CORD_LC_TIMEOUT if lookup timed out
+ * @return CORD_LC_ERR on any other internal error
+ */
+ssize_t cord_lc_res_raw(const cord_lc_rd_t *rd, uint8_t content_format,
+                   char *result, size_t maxlen);
+
+/**
+ * @brief Raw lookup for registered endpoints at a RD server
+ *
+ * @param[in]  rd                the target RD server
+ * @param[in]  content_format    data format wanted from RD server
+ * @param[out] result            buffer holding the raw response
+ * @param[in]  maxlen            max length of @p result
+ *
+ * @return length of result on success
+ * @return CORD_LC_NORSC if there is no endpoint lookup interface at @p rd
+ * @return CORD_LC_TIMEOUT if lookup timed out
+ * @return CORD_LC_ERR on any other internal error
+ */
+ssize_t cord_lc_ep_raw(const cord_lc_rd_t *rd, uint8_t content_format,
+                   char *result, size_t maxlen);
+
+/**
+ * @brief Look for registered resources at a RD server
+ *
+ * @param[in]  rd               the target RD server
+ * @param[out] resources        list of resources at @p rd
+ * @param[in]  limit            max number of resources prepared to receive
+ *
+ * @return number of found resources on success
+ * @return CORD_LC_NORSC if there is no resource lookup interface at @p rd
+ * @return CORD_LC_TIMEOUT if lookup timed out
+ * @return CORD_LC_ERR on any other internal error
+ */
+ssize_t cord_lc_res(const cord_lc_rd_t *rd, cord_lc_res_t *resources, size_t limit);
```

Comment:
This function is resource intensive. I would consider an iterative function on the collected string from coap_lc_res_raw().

```
cord_lc_read_res(const char *source, clif_t *link)
```
where link is the output parameter. However, this function is similar to `coap_decode_links()` from #11189, so not that useful. 

On the other hand, I would not be surprised to see a block based response from the RD. How to incorporate the possibility into the API? Rather than separately call `cord_lc_res_raw()` first, maybe something like this:

```
cord_lc_read_res(const cord_lc_rd_t *rd, char *input, unsigned input_len, clif_t *link)
```
where the cord_lc_rd_t param could be extended to include the block number. That means it could transparently retrieve the next block.

Reply:
> This function is resource intensive. I would consider an iterative function on the collected string from coap_lc_res_raw().

Sorry, I don't quite understand here. What part of the function, do you mean, is resource intensive? Internally, it currently uses _lookup_raw, which are also used by cord_lc_res_raw, and then parses the result and put it in cord_lc_res_t.

I do however find that using iterative function on the collected string is a good idea. As clif_decode_link only decodes one link per call, cord_lc_read_res can be used as a wrapper to it to decode all the links at once and returns it to the user.

> On the other hand, I would not be surprised to see a block based response from the RD.

Block based response would be nice actually. The default GCOAP_PDU_BUF_SIZE is too small even for an RD ep with only 3 resources. Currently, I just set the buffer size bigger but using block option would solve this, I think.

I found #11056 but haven't take a deeper look at it and doesn't really know how to use the API and integrate it here yet, but as you understands it better than me, I'll keep your suggestion in mind ;)

Reply:
> As clif_decode_link only decodes one link per call, cord_lc_read_res can be used as a wrapper to it to decode all the links at once and returns it to the user.

That's the resource intensive part. ;-) By iterative I meant that it would be less intensive for the user to repeatedly call `cord_lc_read_res()` with the output from the query, and receive one clif_t link per call. Maybe the cord_lc_rd_t parameter could remember the last character read from the query result, as well as remember the last block number it retrieved.

Let's wait for clif, and I'd be happy to discuss more.

Reply:
Aah okay. That's what you mean.

> Maybe the cord_lc_rd_t parameter could remember the last character read from the query result, as well as remember the last block number it retrieved.

There are actually [page and count parameters][1] that can be used for the lookup. So this should be easier to implement. Just as a note for now.

[1]: https://tools.ietf.org/html/draft-ietf-core-resource-directory-20#section-6.2

Reply:
Thanks for the link. I actually saw that earlier but thought, "why do we need this when we already have block"? Looking at it more closely, I see that the interface is super simple for a client and eliminates the need for the client to stitch 'sliced' links together from block's fixed size response. On the other hand, if packet space is really limited, block provides certainty in the quantity of data received.

It might actually be worthwhile to ask the core mailing list about the reasoning. If we're concerned about data size, use of CBOR would be valuable here, but I guess that use is orthogonal to how the data is sliced. I also agree with you that we don't need to resolve this issue today.